lib_camera/src/isp/camera_isp_eof_auto_exposure.c

```diff
@@ -0,0 +1,264 @@
+// Copyright 2025 XMOS LIMITED.
+// This Software is subject to the terms of the XMOS Public Licence: Version 1.
+#include <stdint.h>
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+
+#include <print.h>
+#include <xcore/assert.h>
+
+#include "camera_isp.h"
+
+#define HIST_BIN_COUNT      (64)
+#define HIST_QUANT_BITS     (2)
+
+#define AE_MARGIN           (0.1)     // default marging for the auto exposure error
+#define AE_INIT_EXPOSURE    (35)      // initial exposure value
+#define AE_DONE             (0)       // done flag for auto exposure
+
+typedef enum {
+    CHANNEL_RED = 0,
+    CHANNEL_GREEN = 1,
+    CHANNEL_BLUE = 2,
+} channel_order_t;
+
+typedef struct {
+    uint32_t bins[HIST_BIN_COUNT];
+} channel_histogram_t;
+
+typedef struct {
+    uint8_t min;
+    uint8_t max;
+    uint8_t percentile;
+    float skewness;
+    float mean;
+    uint32_t max_count;
+    uint32_t min_count;
+    uint32_t per_count;
+} channel_stats_t;
+
+typedef struct {
+    channel_histogram_t histogram_red;
+    channel_histogram_t histogram_green;
+    channel_histogram_t histogram_blue;
+} histograms_t;
+
+typedef struct {
+    channel_stats_t stats_red;
+    channel_stats_t stats_green;
+    channel_stats_t stats_blue;
+} statistics_t;
+
+
+static
+void stats_simple(
+    channel_histogram_t* histogram,
+    channel_stats_t* stats,
+    const float inv_img_size)
+{
+    // Calculate the histogram
+    uint8_t temp_min = 0;
+    uint8_t temp_max = 0;
+    float temp_mean = 0;
+
+    for (int k = 0; k < HIST_BIN_COUNT; k++) {
+        uint32_t bin_count = histogram->bins[k];
+        // mean
+        temp_mean += bin_count * k;
+        // max and min
+        if (bin_count != 0) {  // the last that is not zero
+            temp_max = k;
+            if (temp_min == 0) { // first time is zero, then min is set
+                temp_min = k;
+            }
+        }
+    }
+    // max and min count
+    stats->max_count = histogram->bins[temp_max];
+    stats->min_count = histogram->bins[temp_min];
+
+    // biased downwards due to truncation
+    stats->max = (temp_max << HIST_QUANT_BITS);
+    stats->min = (temp_min << HIST_QUANT_BITS);
+    stats->mean = (temp_mean) * (1 << HIST_QUANT_BITS) * inv_img_size;
+}
+
+static
+void stats_skewness(
+    channel_histogram_t* histogram,
+    channel_stats_t* stats,
+    const float inv_img_size)
+{
+    const float zk_values[] = {
+      -1.000000, -0.907753, -0.821362, -0.740633, -0.665375, -0.595396,
+      -0.530504, -0.470508, -0.415214, -0.364431, -0.317968, -0.275632,
+      -0.237231, -0.202574, -0.171468, -0.143721, -0.119142, -0.097538,
+      -0.078717, -0.062488, -0.048659, -0.037037, -0.027431, -0.019648,
+      -0.013497, -0.008786, -0.005323, -0.002915, -0.001372, -0.000005,
+      -0.000108,    -4e-06,     4e-06,  0.000108,    0.0005,  0.001372,

```

Comment:
The positive and negative values in this list are symmetric with the exception of the entries for -0.000005 and 0.0005.  This may be an inadvertent mistake.